### ISSUE TITLE
Improve dynamic module loading

### ISSUE THREAD

**Issue by ajfriend** [uber/h3-py#80]
As mentioned here, it would be nice to find a less-hacky solution for dynamically creating the different APIs.

https://github.com/uber/h3-py/blob/963241447bba1151958199c6ac30d726c34a2345/src/h3/api/_api_template.py#L6-L18

Perhaps using:
- https://docs.python.org/3.8/library/functions.html#__import__
- import hooks: https://www.python.org/dev/peps/pep-0302/

Couple of other, potentially helpful references:
- [Dependency Injection with Import Hooks in Python 3
](https://dev.to/dangerontheranger/dependency-injection-with-import-hooks-in-python-3-5hap)
- [Dynamic Import in Python 3
](https://www.bnmetrics.com/blog/dynamic-import-in-python3)

**Comment by dfellis** [comment]
I would recommend re-evaluating the need to have dynamic imports in the first place. Hard-to-debug issues can crop up from that and it can confuse automatic tooling in IDEs.

If there's no easy way to do the imports non-dynamically, perhaps a build script to automatically generate a file defining the imports that can be checked into the repository would make more sense?

**Comment by ajfriend** [comment]
Agreed that it's a trade-off to be considered carefully. I don't want it to be *too* magical.

The motivation was that I couldn't see another way to do it without a lot of cut-and-paste when producing the different APIs. I figured that having a single, shared code path between all APIs would probably be less bug-prone in the long run. One benefit, for example, is that a test for `set_str` runs basically the same code path as `array_int`; we just have to check that the conversion functions for each API work correctly. Without the code de-duplication, we'd need (basically duplicated) tests for each API.

That all said, I'm not entirely happy with how the current system works, and am more than open to suggestions for nice solutions!

**Comment by ajfriend** [comment]
More random links about dynamically creating modules for the APIs:

- https://learning.oreilly.com/library/view/python-cookbook/0596001673/ch15s03.html
- https://www.pythondoeswhat.com/2011/12/easily-import-dynamically-created.html?m=1
